crucible-llvm/src/Lang/Crucible/LLVM/SymIO.hs

```diff
@@ -0,0 +1,560 @@
+        args (getOverrideArgs >>= \args' -> cont Nothing args') Nothing
+    W4P.Unassigned -> cont Nothing args
+
+
+-- | Allocate a fresh (integer/bitvector(32)) file descriptor that is associated with the given 'SymIO.FileHandle'
+--
+-- NOTE that this is a file descriptor in the POSIX sense, rather than a @FILE*@
+-- or the underlying 'SymIO.FileHandle'.
+--
+-- NOTE that we truncate the file descriptor source to 32 bits in this function;
+-- it could in theory overflow.
+--
+-- NOTE that the file descriptor counter is incremented monotonically as the
+-- simulator hits calls to @open@; this means that calls to @open@ in parallel
+-- control flow branches would get sequential file descriptor values whereas the
+-- real program would likely allocate the same file descriptor value on both
+-- branches. This could be relevant for some bug finding scenarios.
+--
+-- TODO It would be interesting if we could add a symbolic offset to these
+-- values so that we can't make any concrete assertions about them. It isn't
+-- clear if that ever happens in real code. If we do that, we need an escape
+-- hatch to let us allocate file descriptors 0, 1, and 2 if needed.
+allocateFileDescriptor
+  :: (IsSymInterface sym, HasPtrWidth wptr)
+  => LLVMFileSystem wptr
+  -> SymIO.FileHandle sym wptr
+  -> OverrideSim p sym ext r args ret (W4.SymBV sym 32)
+allocateFileDescriptor fsVars fh = do
+  sym <- getSymInterface
+  modifyGlobal (llvmFileDescMap fsVars) $ \(FDescMap next descMap) -> do
+    fdesc <-  liftIO $ W4.bvLit sym (PN.knownNat @32) (BVS.mkBV (PN.knownNat @32) (toInteger next))
+    let ptrMap' = Map.insert next (W4P.justPartExpr sym fh) descMap
+    return (fdesc, FDescMap (next + 1) ptrMap')
+
+loadFileIdent
+  :: (IsSymInterface sym, HasLLVMAnn sym, HasPtrWidth wptr)
+  => GlobalVar Mem
+  -> LLVMPtr sym wptr
+  -> OverrideSim p sym ext r args ret (SymIO.FileIdent sym)
+loadFileIdent memOps filename_ptr = do
+  mem <- readGlobal memOps
+  sym <- getSymInterface
+  filename_bytes <- liftIO $ loadString sym mem filename_ptr Nothing
+  liftIO $ W4.stringLit sym (W4.Char8Literal (BS.pack filename_bytes))
+
+returnIOError32
+  :: IsSymInterface sym
+  => OverrideSim p sym ext r args ret (W4.SymBV sym 32)
+returnIOError32 = do
+  sym <- getSymInterface
+  liftIO $ W4.bvLit sym (PN.knownNat @32) (BVS.mkBV (PN.knownNat @32) (-1))
+
+returnIOError
+  :: forall wptr p sym ext r args ret
+  . (IsSymInterface sym, HasPtrWidth wptr)
+  => OverrideSim p sym ext r args ret (W4.SymBV sym wptr)
+returnIOError = do
+  sym <- getSymInterface
+  liftIO $ W4.bvLit sym PtrWidth (BVS.mkBV PtrWidth (-1))
+
+openFile
+  :: (IsSymInterface sym, HasLLVMAnn sym, HasPtrWidth wptr)
+  => LLVMFileSystem wptr
+  -> LLVMOverride p sym
+           (EmptyCtx ::> LLVMPointerType wptr
+                     ::> BVType 32)
+           (BVType 32)
+openFile fsVars =
+  [llvmOvr| i32 @open( i8*, i32 ) |]
+  -- TODO add mode support by making this a varargs function
```

Comment:
I don't remember seeing documentation about which flags and such are supported.  Probably worth writing that down.

Reply:
I added documentation on that to the README for crux-llvm. Do you want the errata duplicated in this module?

Reply:
No, better to have it in one place, I think.  Maybe reference it here? I'm indifferent.